source/dofs/dof_tools.cc

```diff
@@ -492,26 +444,15 @@ namespace DoFTools
 
 
 
-  template <int dim, int spacedim>
-  void
-  extract_dofs(const DoFHandler<dim, spacedim> &dof,
-               const BlockMask &                block_mask,
-               std::vector<bool> &              selected_dofs)
-  {
-    // simply forward to the function that works based on a component mask
-    extract_dofs(dof, dof.get_fe().component_mask(block_mask), selected_dofs);
-  }
-
-
-
-  template <int dim, int spacedim>
+  template <int dim, int spacedim, typename DoFHandlerType>
   void
-  extract_dofs(const hp::DoFHandler<dim, spacedim> &dof,
-               const BlockMask &                    block_mask,
-               std::vector<bool> &                  selected_dofs)
+  extract_dofs(const DoFHandlerType &dof,
+               const BlockMask &     block_mask,
+               std::vector<bool> &   selected_dofs)
   {
     // simply forward to the function that works based on a component mask
-    extract_dofs(dof, dof.get_fe().component_mask(block_mask), selected_dofs);
+    extract_dofs<dim, spacedim, DoFHandlerType>(
```

Comment:
You can use `DoFHandlerType::dimension` and `DoFHandlerType::space_dimension` here and in the other places.

Reply:
Thanks for this valuable hint! Indeed this makes life so  much easier.
I have (again...) a new proposal now and could merge all the ideas I had at the beginning. Let me explain in a nutshell what I tried to achieve and I begin with the testsuite (and I like it really very much):
Basically I splitted the 12a test into a dof_tools_common_parallel.h file, which is similar to the non parallel version, just with the parallel triangulation and not using (the non supported) 1d elements. So basically if someone wants to make a test for MPI in the future, he can just include this file, similar to dof_tools_common.
Further I made the check_this() template dependent of the DoFHandler type - also this means to change many code lines. The benefit is, that dof_cools_common.h can now test for both DoFHandlers (hp and not hp) without code deduplicating. I demonstrate this with the updated test dof_tools_12: it's basically the same code as before, but now tests against the normal DoFHandler and the hp::DoFHandler. The other tests could just do the same, if every function implemented the full functionality for DoFHandler and hp::DoFHandler. Until this they just have to include the dof_tools_fake_hp.inst.in so that basically just the normal DoFHandler is being tested. So in my opinion this will make testing in the future easier, when more hp::DoFHandler stuff gets merged (maybe some tests already work with the hp::DoFHandler, I didn't really look into this).

Now to the extract_dofs() function: as you proposed I changed it, that it doesn't depend on dim or spacedim template arguments and therefore auto deduction works as expected.

I still feature the problem with the different output files on my local machine and some reference output files, what I describe above - I really don't know where the different outputs come from and I don't know, why the whitespaces aren't there in the reference output files. This could be a problem with the output file of test 12a, since it includes the whitespaces in the FESystems, also the updated test 12 output files from my system include these whitespaces.
Greetings,
Mathias